src/NuGet.Services.V3/CommitCollectorUtility.cs

```diff
@@ -0,0 +1,170 @@
+    {
+        private readonly ICatalogClient _catalogClient;
+        private readonly IV3TelemetryService _telemetryService;
+        private readonly IOptionsSnapshot<ICommitCollectorConfiguration> _options;
+        private readonly ILogger<CommitCollectorUtility> _logger;
+
+        public CommitCollectorUtility(
+            ICatalogClient catalogClient,
+            IV3TelemetryService telemetryService,
+            IOptionsSnapshot<ICommitCollectorConfiguration> options,
+            ILogger<CommitCollectorUtility> logger)
+        {
+            _catalogClient = catalogClient ?? throw new ArgumentNullException(nameof(catalogClient));
+            _telemetryService = telemetryService ?? throw new ArgumentNullException(nameof(telemetryService));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            if (_options.Value.MaxConcurrentCatalogLeafDownloads <= 0)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(options),
+                    $"The {nameof(ICommitCollectorConfiguration.MaxConcurrentCatalogLeafDownloads)} must be greater than zero.");
+            }
+        }
+
+        public IEnumerable<CatalogCommitItemBatch> CreateSingleBatch(IEnumerable<CatalogCommitItem> catalogItems)
+        {
+            if (!catalogItems.Any())
+            {
+                return Enumerable.Empty<CatalogCommitItemBatch>();
+            }
+
+            var maxCommitTimestamp = catalogItems.Max(x => x.CommitTimeStamp);
+
+            return new[]
+            {
+                new CatalogCommitItemBatch(
+                    catalogItems,
+                    key: null,
+                    commitTimestamp: maxCommitTimestamp),
+            };
+        }
+
+        public List<CatalogCommitItem> GetLatestPerIdentity(IEnumerable<CatalogCommitItem> items)
+        {
+            return items
+                .GroupBy(x => x.PackageIdentity)
+                .Select(GetLatestPerSingleIdentity)
+                .ToList();
+        }
+
+        private CatalogCommitItem GetLatestPerSingleIdentity(IEnumerable<CatalogCommitItem> entries)
+        {
+            CatalogCommitItem max = null;
+            foreach (var entry in entries)
+            {
+                if (max == null)
+                {
+                    max = entry;
+                    continue;
+                }
+
+                if (!StringComparer.OrdinalIgnoreCase.Equals(max.PackageIdentity, entry.PackageIdentity))
+                {
+                    throw new InvalidOperationException("The entries compared should have the same package identity.");
+                }
+
+                if (entry.CommitTimeStamp > max.CommitTimeStamp)
+                {
+                    max = entry;
+                }
+                else if (entry.CommitTimeStamp == max.CommitTimeStamp)
```

Comment:
Super nit, consider reducing nesting:

```csharp
if (entry.CommitTimeStamp == max.CommitTimeStamp)
{
    ...
}

if (entry.CommitTimeStamp > max.CommitTimeStamp)
{
    max = entry;
}
```

Reply:
Chose not to do this to minimize code changes while moving code.